tokio/src/io/poll_evented.rs

```diff
@@ -151,16 +151,38 @@ feature! {
         {
             use std::io::Read;
 
-            let n = ready!(self.registration.poll_read_io(cx, || {
+            let mut n = 0;
+
+            let res = self.registration.poll_read_io(cx, || {
                 let b = &mut *(buf.unfilled_mut() as *mut [std::mem::MaybeUninit<u8>] as *mut [u8]);
-                self.io.as_ref().unwrap().read(b)
-            }))?;
-
-            // Safety: We trust `TcpStream::read` to have filled up `n` bytes in the
-            // buffer.
-            buf.assume_init(n);
-            buf.advance(n);
-            Poll::Ready(Ok(()))
+                let len = b.len();
+
+                // n is the useful indicator of how many bytes are read, not `res`
+                n = self.io.as_ref().unwrap().read(b)?;
+
+                if n == 0 || n == len {
+                    // if we return a full buffer, the socket buffer might not be drained
+                    Ok(n)
+                } else {
+                    // if we read a partially full buffer, this is sufficient to show that the
+                    // socket buffer has been drained
+                    Err(io::ErrorKind::WouldBlock.into())
+                }
+            });
+
+            if let Poll::Ready(Err(e)) = res {
```

Comment:
I'm assuming that `Poll::Pending` is returned if the closure in `poll_read_io` returns `WouldBlock`?

Reply:
Yes, `poll_read_io` will return `Poll::Pending` if the closure returns `WouldBlock`, so this `if let` does _not_ trigger in the case of `WouldBlock`. The surrounding function would then return `Poll::Ready` since the `if n > 0` check succeeds if a partial read happened.

I note that since `poll_read_io` returns `Poll::Pending`, it will have registered the waker for a wakeup when it becomes ready again. However, if we then proceed to return `Poll::Ready`, then this means that the waker was registered unnecessarily. This is not _incorrect_ as unnecessary wakeups are always allowed, but we might want to consider the optimization of not registering the waker in this case.